sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,256 @@
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn()
+            .context("failed to spawn openocd")?;
+        self.openocd_server_process.replace(Some(child));
+
+        // Give OpenOCD time to start up
+        thread::sleep(Duration::from_millis(1000));
+
+        log::info!("Connecting to OpenOCD tcl interface...");
+
+        let stream = TcpStream::connect(format!("localhost:{OPENOCD_TCL_PORT}"))
+            .context("failed to connect to openocd telnet interface")?;
+        self.openocd_socket_stream.set(Some(stream));
+
+        // Test the connection by asking for OpenOCD's version
+        let version = self
+            .send_tcl_cmd("version")
+            .context("failed to get OpenOCD version")?;
+        log::info!("OpenOCD version: {version}");
+
+        Ok(())
+    }
+
+    fn stop(&self) -> Result<()> {
+        log::info!("Stopping OpenOCD...");
+
+        let Some(mut child) = self.openocd_server_process.take() else {
+            // OpenOCD wasn't started, do nothing
+            return Ok(());
+        };
+
+        // TODO: send "shutdown" command and wait before killing the process
+        Ok(child.kill()?)
+    }
+
+    /// Send a TCL command to OpenOCD and wait for its response.
+    fn send_tcl_cmd(&self, cmd: &str) -> Result<String> {
+        // Take the stream. The stream will not be replaced on communication
+        // errors, causing future commands to also fail.
+        let Some(mut stream) = self.openocd_socket_stream.take() else {
+            return Err(OpenOcdServerError::OpenOcdNotRunning.into());
+        };
+
+        // The protocol is to send the command followed by a `0x1a` byte,
+        // see https://openocd.org/doc/html/Tcl-Scripting-API.html#Tcl-RPC-server
+        stream
+            .write_all(cmd.as_bytes())
+            .context("failed to send a command to OpenOCD server")?;
+        stream
+            .write_all(&[0x1a])
+            .context("failed to send a command to OpenOCD server")?;
+
+        stream.flush().context("failed to flush stream")?;
+
+        let answer = stream
+            .borrow_mut()
+            .bytes()
+            .take_while(|x| match x {
+                Ok(x) => *x != 0x1a,
+                _ => false,
+            })
+            .collect::<Result<_, _>>()
+            .context("failed to read the result of a command from OpenOCD server")?;
+
+        self.openocd_socket_stream.replace(Some(stream));
+        String::from_utf8(answer).context("failed to parse OpenOCD response as UTF-8")
+    }
+}
+
+impl Drop for OpenOcdServer {
+    fn drop(&mut self) {
+        self.stop().ok();
+    }
+}
+
+impl Jtag for OpenOcdServer {
+    fn connect(&self, tap: JtagTap) -> Result<()> {
+        self.start(tap)
+    }
+
+    fn disconnect(&self) -> Result<()> {
+        self.stop()
+    }
+
+    fn read_lc_ctrl_reg(&self, reg: &LcCtrlReg) -> Result<u32> {
+        let reg_offset = reg.word_offset();
```

Comment:
probably want to fail early if the LC TAP isn't selected, i.e., if we have selected the CPU TAP right?

Reply:
same for the write function below

Reply:
ok to do in a follow up.